guppylang/checker/cfg_checker.py

```diff
@@ -260,13 +266,43 @@ def check_bb(
     return checked_bb
 
 
-def check_rows_match(row1: Row[Variable], row2: Row[Variable], bb: BB) -> None:
+def check_rows_match(
+    row1: Row[Variable], row2: Row[Variable], bb: BB, globals: Globals
+) -> None:
     """Checks that the types of two rows match up.
 
     Otherwise, an error is thrown, alerting the user that a variable has different
     types on different control-flow paths.
     """
     map1, map2 = {v.name: v for v in row1}, {v.name: v for v in row2}
+
+    # If block signature lengths don't match but no undefined error was thrown, some
+    # variables may be shadowing global variables.
+    if map1.keys() != map2.keys():
+        for x in map1.keys() ^ map2.keys():
+            if x in globals:
+                v = map1.get(x) or map2.get(x)
+                assert v is not None
+                g = globals[x]
+                assert hasattr(g, "ty")
+                if v.ty != g.ty:
+                    ident = (
+                        "Expression"
+                        if v.name.startswith("%")
+                        else f"Variable `{v.name}`"
+                    )
+                    use = bb.containing_cfg.live_before[bb][v.name].vars.used[v.name]
+                    err = BranchTypeError(use, ident)
+                    err.add_sub_diagnostic(BranchTypeError.TypeHint(v.defined_at, v.ty))
+                    # We don't add a location to the type hint for the global variable,
+                    # since  it could lead to cross-file diagnostics (which are not
+                    # supported) or refer to long function definitions.
+                    err.add_sub_diagnostic(BranchTypeError.GlobalHint(None, g.ty))
+                    raise GuppyError(err)
+                # Remove variables from comparison below if types match.
+                map1.pop(x, None)
+                map2.pop(x, None)
+
     assert map1.keys() == map2.keys()
     for x in map1:
```

Comment:
It might be nicer to move this logic into the existing loop to avoid code duplication? Maybe something along the lines of

```python
for x in map1.keys() | map2.keys():
    v1 = map1.get(x) or globals[x]
    v2 = map2.get(x) or globals[x]
    if v1.ty != v2.ty:
        # ...
        err = BranchTypeError(use, ident)
        sub1 = (
            BranchTypeError.TypeHint(v1.defined_at, v1.ty) if isinstance(v1, Variable)
            else BranchTypeError.GlobalHint(None, v1.ty)
        )
        sub2 = (
            BranchTypeError.TypeHint(v2.defined_at, v2.ty) if isinstance(v2, Variable)
            else BranchTypeError.GlobalHint(None, v2.ty)
        )
        err.add_sub_diagnostic(sub1)
        err.add_sub_diagnostic(sub2)
        raise GuppyError(err)
```